sdks/common/runner-api/src/main/proto/beam_runner_api.proto

```diff
@@ -0,0 +1,436 @@
+// between PCollections.
+message PTransform {
+
+  // (Required) A pipeline-scoped unique id for this node.
+  int64 id = 1;
+
+  // (Required) A unique program-readable name for the application node.
+  //
+  // This should be stable over multiple runs of a driver program for the
+  // purposes of logging and associating pipeline state with a node.
+  string label = 2;
+
+  // (Optional) A URN that identifies what type of transform this is.
+  //
+  // If absent, this must be an "anonymous" composite transform.
+  //
+  // If present, this URN should suffice to understand the payload and
+  // the two together should fully specify the SDK-independent and
+  // runner-independent semantics of the transform.
+  string urn = 3;
+
+  // (Optional) An arbitrary payload whose schema is determined by the URN.
+  //
+  // For primitive transform in the Runner API, the payload is well defined:
+  //
+  //  - when the URN is "urn:beam:transforms:pardo" it is a ParDoPayload
+  //  - when the URN is "urn:beam:transforms:read" it is a ReadPayload
+  //  - when the URN is "urn:beam:transforms:gbk" it is a GroupByKeyPayload
+  //  - when the URN is "urn:beam:transforms:window" it is a WindowPayload
+  //  - when the URN is "urn:beam:transforms:flatten" it is absent
+  //
+  // For some special composite transforms, the payload is also well-defined:
+  //
+  //  - when the URN is "urn:beam:transforms:combine" it is a CombinePayload
+  //
+  // For other composites, the payload will often be absent.
+  //
+  // TODO: duplicate PTransform's fields into each Payload to save indirection?
+  google.protobuf.Any payload = 4;
+
+  // (Optional) Static display data.
+  map<string, string> display_data = 6;
+}
+
+// The payload for the primitive ParDo transform.
+message ParDoPayload {
+
+  // (Required) The DoFn for this ParDo.
+  FunctionSpec do_fn = 1;
+
+  // (Required) Whether the DoFn requires access to the current window (this
+  // may force runners to perform additional work prior to executing this ParDo)
+  //
+  // TODO: consider a DoFnPayload that holds these facts common to all DoFns
+  bool requires_window_access = 2;
+
+  // (Optional) A map from local input name to the specification for how to
+  // materialize and read it as a side input
+  map<string, SideInput> side_inputs = 3;
+
+  // (Optional) if the DoFn uses state, a map from id to the spec for the
+  // state cell.
+  map<string, StateSpec> state_specs = 4;
+
+  // (Optional) if the DoFn uses timers, a map from id to the spec for the
+  // timer.
+  map<string, TimerSpec> timer_specs = 5;
+}
+
+message StateSpec {
+  // TODO: AST for state spec
+}
+
+message TimerSpec {
+  // TODO: AST for timer spec
+}
+
+// The payload for the primitive Read transform.
+//
+// TODO: audit what needs to be in this payload
+message ReadPayload {
+
+  // (Required) The source for this Read
+  FunctionSpec source = 1;
+
+  // (Required) Whether the source is bounded or unbounded
```

Comment:
Putting it on the PCollection. The SDK can do the analysis (as it does with e.g. propagating windowing strategy)